formula-coroutines/src/main/java/com/instacart/formula/coroutines/FlowRuntime.kt

```diff
@@ -0,0 +1,42 @@
+package com.instacart.formula.coroutines
+
+import com.instacart.formula.FormulaRuntime
+import com.instacart.formula.IFormula
+import com.instacart.formula.internal.ThreadChecker
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.channels.trySendBlocking
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+
+object FlowRuntime {
+    fun <Input : Any, Output : Any> start(
+        input: Flow<Input>,
+        formula: IFormula<Input, Output>
+    ): Flow<Output> {
+        val threadChecker = ThreadChecker()
+        return callbackFlow<Output> {
+            threadChecker.check("Need to subscribe on main thread.")
+
+            var runtime = FormulaRuntime(threadChecker, formula, ::trySendBlocking, channel::close)
+
+            input.collect { input ->
```

Comment:
TODO: Using `collect` here is a bug, need to fix it

Reply:
would you mind expanding a  bit on why using `collect` is a bug here?

Reply:
Is it because currently there is no way to cancel it? Meaning there is no job instance that would allow us to dispose of it?

Reply:
Not exactly, it's actually hard to see it if you aren't very familiar with coroutines (I'm still pretty new so this took me a bit to figure out). I found this bug as I was adding more tests to the coroutine runtime. The issue came up when I added a test for formula termination. 

The problem here is that `collect` is a suspending operation and it will block execution until the flow is exhausted. When you pass `MutableSharedFlow`, it will never exhaust itself so nothing below `input.collect{}` will run including `awaitClose`.

To fix this, I replaced `collect` with
```kotlin
input.onEach {}.launchIn(this)
awaitClose {}
```

Reply:
Thanks for the explanation. You are exactly right, those hot Flows never finish. I was going to propose what you suggest as a solution as well.

The question I have about that is, do we need to call job.cancel() on the job returned by launchIn() inside awaitClose inside the if?

Reply:
I don't think we need to do that as we pass our `scope` to `launchIn(this)` so it should be canceled once the `callbackFlow` is canceled.